bout-particle-push.cxx

```diff
@@ -0,0 +1,767 @@
+  //   }
+  // }
+
+  // PetscInterface::label_dmplex_edges(dm, PetscInterface::face_sets_label,
+  //                                    vertex_starts, vertex_ends, edge_labels);
+
+  // save a HDF5 file containing the DM for diagnostics
+  PetscViewer viewer;
+  // Set a name for the DMPlex object (important for HDF5)
+  PetscObjectSetName((PetscObject)dm, "hypnotoad_dmplex_mesh");
+  // Create an HDF5 viewer
+  PetscViewerHDF5Open(PETSC_COMM_WORLD, "hypnotoad_dmplex_mesh_output.h5", FILE_MODE_WRITE, &viewer);
+  // Set viewer format to PETSC_VIEWER_HDF5_PETSC for compatibility
+  PetscViewerPushFormat(viewer, PETSC_VIEWER_HDF5_PETSC);
+  // Save the DMPlex to the HDF5 file
+  DMView(dm, viewer);
+  // Clean up
+  PetscViewerDestroy(&viewer);
+  output << "Finished DMPlex creation and diagnostic \n";
+  output << "Begin particle push \n";
+  // get data from BOUT for forces in particle push
+  Field2D phi{mesh};
+  Field2D ex{mesh};
+  Field2D ey{mesh};
+  auto& opt = Options::root();
+  phi = opt["mesh"]["phi"].as<Field2D>();
+  ex = opt["mesh"]["ex"].as<Field2D>();
+  ey = opt["mesh"]["ey"].as<Field2D>();
+
+  /*
+   *
+   *
+   *
+   *
+   *
+   * Below here we setup basic advection on the dmplex
+   *
+   *
+   *
+   *
+   *
+   *
+   */
+  {
+    const int ndim = 2;
+    const int npart_per_cell = Options::root()["neso_particles"]["npart_per_cell"].withDefault(1);
+    const REAL dt = Options::root()["neso_particles"]["dt"].withDefault(0.01);
+    const int nsteps = Options::root()["neso_particles"]["nsteps"].withDefault(10);
+    Field2D density{mesh};
+    mesh->get(density, "density", 0.0, false);
+    mesh->communicate(density);
+
+    // Create a mesh interface from the DM
+    auto neso_mesh = std::make_shared<PetscInterface::DMPlexInterface>(
+        dm, 0, MPI_COMM_WORLD);
+    // Create a mapper for mapping particles into cells.
+    auto mapper =
+        std::make_shared<PetscInterface::DMPlexLocalMapper>(sycl_target, neso_mesh);
+    // Create a domain from the neso_mesh and the mapper.
+    auto domain = std::make_shared<Domain>(neso_mesh, mapper);
+
+    // Create the particle properties (note that if you are using the Reactions
+    // project it has its owne particle spec builder).
+    ParticleSpec particle_spec{ParticleProp(Sym<REAL>("P"), ndim, true),
+                               ParticleProp(Sym<REAL>("V"), ndim),
+                               ParticleProp(Sym<REAL>("F"), ndim),
+                               ParticleProp(Sym<REAL>("Q"), 1),
+                               ParticleProp(Sym<REAL>("TSP"), 2),
+                               ParticleProp(Sym<INT>("CELL_ID"), 1, true),
+                               ParticleProp(Sym<INT>("ID"), 1)};
+
+    // Create a Particle group with our specied particle properties.
+    auto A =
+        std::make_shared<ParticleGroup>(domain, particle_spec, sycl_target);
+
+    // Create some particle data
+
+    std::mt19937 rng_pos(52234234 + mpi_rank);
```

Comment:
This seed should have a better creation mechanism at some point.

Reply:
Maybe retrieve one from the config file and sample a random one as the default if the user did not provide one?